### PR TITLE
Implementation of r2_score function

### PR DESCRIPTION
**Pull request by mottodora**
I implemented the r2_score function that calculates R^2(coefficient of determination) regression score function.
As of now, the code is compatible with scikit-learn, only when the "sample_weight" argument is set to "none". 
Is this argument necessary for chainer? If necessary, I will implement the changes.
Additionally, I would like to know whether the type of this argument is Variable or ndarray?

### CODE REVIEW

**Comment by delta2323** [comment]
Please fix violation of the coding guideline by referencing [travis result](https://travis-ci.org/pfnet/chainer/jobs/176959744).

File: chainer/functions/evaluation/r2_score.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        type_check.expect(in_types.size() == 2)
+        pred_type, true_type = in_types
+
+        type_check.expect(
+            pred_type.dtype.kind == 'f',
+            true_type.dtype.kind == 'f'
+        )
+
+        type_check.expect(
+            pred_type.ndim >= true_type.ndim,
```
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: We do not need this condition because `pred_type.shape == true_type.shape` implies the equality of `ndim`'s

File: chainer/functions/evaluation/r2_score.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    def forward(self, inputs):
+        xp = cuda.get_array_module(*inputs)
+        pred, true = inputs
+        SS_res = xp.sum((pred-true)**2, axis=0)
+        SS_tot = xp.sum((true-xp.mean(true, axis=0))**2, axis=0)
+        if self.multioutput == 'uniform_average':
+            return xp.asarray((1 - SS_res / SS_tot).mean(), dtype=pred.dtype),
+        elif self.multioutput == 'raw_values':
+            return xp.asarray((1 - SS_res / SS_tot), dtype=pred.dtype),
```
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: scikit learn seems to set r2 value to 0 when `SS_tot` is 0.

File: chainer/functions/evaluation/r2_score.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def r2_score(pred, true, sample_weight=None, multioutput='uniform_average'):
+    """Computes R^2(coefficient of determination) regression score function.
+
+    Args:
+        pred(Variable): Variable holding a vector or matrix of estimated \
```
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: We do not need backslashes in breaking the description of an argument.

File: chainer/functions/evaluation/r2_score.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    Returns:
+        Variable: A Variable holding a scalar array of the R^2 score if \
+                'multioutput' is 'uniform_average' or a vector of R^2 \
+                scores if 'multioutput' is 'raw_values'.
+
+    .. note:: This function is non-differentiable.
+
+    """
+    return R2_score(sample_weight=sample_weight, multioutput=multioutput)\
+            (pred, true)
```
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: You can break the line in the middle of parentheses as follows:
```python
return R2_score(sample_weight=sample_weight,
                multioutput=multioutput)(pred, true)
```

File: chainer/functions/evaluation/r2_score.py (hunk 3 above)
Comment (mottodora) [inline chainer/functions/evaluation/r2_score.py]: I deleted backslash.

File: chainer/functions/evaluation/r2_score.py (hunk 4 above)
Comment (mottodora) [inline chainer/functions/evaluation/r2_score.py]: I fixed this line.

File: chainer/functions/evaluation/r2_score.py [hunk 5]
```diff
@@ -22,41 +23,48 @@ def check_type_forward(self, in_types):
...
 
 
 def r2_score(pred, true, sample_weight=None, multioutput='uniform_average'):
     """Computes R^2(coefficient of determination) regression score function.
 
     Args:
-        pred(Variable): Variable holding a vector or matrix of estimated \
+        pred(Variable): Variable holding a vector or matrix of estimated
                 target values.
         true(Variable): Variable holding a vector or matrix of correct target \
```
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: Please remove backslash here, too.

File: chainer/functions/evaluation/r2_score.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        SS_res = xp.sum((pred - true) ** 2, axis=0)
+        SS_tot = xp.sum((true - xp.mean(true, axis=0)) ** 2, axis=0)
+        if self.multioutput == 'uniform_average':
+            if xp.any(SS_tot == 0):
+                return xp.asarray(0.0, dtype=pred.dtype),
+            else:
+                return xp.asarray((1 - SS_res / SS_tot).mean(),
+                                  dtype=pred.dtype),
+        elif self.multioutput == 'raw_values':
+            if xp.any(SS_tot == 0):
```
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: If the reason for this branching is to avoid zero division by `SS_tot`, we cannot avoid it because `SS_res / SS_tot` is evaluated anyway in case `SS_tot == 0`. Are there other reason for this branching? If not, I recommend to remove the branching. By removing it, we can expect faster forward computation.

File: chainer/functions/evaluation/r2_score.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                return xp.where(SS_tot != 0, 1 - SS_res / SS_tot, 0.0)\
+                    .astype(pred.dtype),
+            else:
+                return xp.asarray((1 - SS_res / SS_tot), dtype=pred.dtype),
+
```
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: We can merge `uniform_average` case and `raw_values` case as follows

```python
ret = xp.asarray((1 - SS_res / SS_tot), dtype=pred.dtype)
if self.multioutput == `uniform_average`
    return ret.mean()
else:
    return ret
```

File: chainer/functions/evaluation/r2_score.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+def r2_score(pred, true, sample_weight=None, multioutput='uniform_average'):
+    """Computes R^2(coefficient of determination) regression score function.
+
+    Args:
+        pred(Variable): Variable holding a vector or matrix of estimated
+                target values.
```
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: `pred` can be not only a vector or a matrix but a tensor of any dimension.

File: chainer/functions/evaluation/r2_score.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        true(Variable): Variable holding a vector or matrix of correct target \
```
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: `true` can be not only a vector or a matrix but a tensor of any dimension.

File: chainer/functions/evaluation/r2_score.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+                values.
+        sample_weight: None.
```
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: How about writing the description in more detail like this:

Current implementation allow this argument to be `None` only.

File: chainer/functions/evaluation/r2_score.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        multioutput(string): ['uniform_average', 'raw_values']. if
+                'uniform_average', this function return an average of R^2
```
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: return*s*

File: chainer/functions/evaluation/r2_score.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+                score of multiple output. If 'raw_average', this function
+                return a set of R^2 score of multiple output.
+    Returns:
+        Variable: A Variable holding a scalar array of the R^2 score if
+                'multioutput' is 'uniform_average' or a vector of R^2
+                scores if 'multioutput' is 'raw_values'.
+
+    .. note:: This function is non-differentiable.
```
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: As forward computation of this function consists of differentiable operations, we can back propagate an error if we implement `backward` method appropriately. Do you think it is beneficial to implement backward?

File: tests/chainer_tests/functions_tests/evaluation_tests/test_r2_score.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        self.assertEqual(y.data.dtype, self.dtype)
+        if self.multioutput == 'uniform_average':
+            self.assertEqual((), y.data.shape)
+        elif self.multioutput == 'raw_values':
+            self.assertEqual(x_data.shape[1:], y.data.shape)
+
+        expected = r2_score(self.x, self.t, sample_weight=None,
+                            multioutput=self.multioutput)
+        testing.assert_allclose(
+            expected, cuda.to_cpu(y.data), **self.check_forward_options)
```
Comment (delta2323) [inline tests/chainer_tests/functions_tests/evaluation_tests/test_r2_score.py]: You need not transfer `y.data` to CPU because `testing.assert_allclose` does it.

File: chainer/functions/evaluation/r2_score.py (hunk 6 above)
Comment (mottodora) [inline chainer/functions/evaluation/r2_score.py]: scikit learn returns 0 when SS_tot is 0. If not brancing, this function will return NaN.

File: chainer/functions/evaluation/r2_score.py (hunk 12 above)
Comment (mottodora) [inline chainer/functions/evaluation/r2_score.py]: I think that there is no difference between mean squared error and r_squared as loss function.

File: chainer/functions/evaluation/r2_score.py (hunk 6 above)
Comment (delta2323) [inline chainer/functions/evaluation/r2_score.py]: I thought something like this:

```python
ret = xp.where(SS_tot != 0, 1 - SS_res / SS_tot, 0.0).astype(pred.dtype)
if self.multioutput == 'uniform_average':
    return ret.mean()
else:
    return ret
```

Does it make sense to you?

Comment (mottodora) [inline chainer/functions/evaluation/r2_score.py]: Yes. I'm going to implement like this.

**Comment by delta2323** [comment]
Could you merge the latest master branch?

**Comment by delta2323** [comment]
LGTM

**Comment by delta2323** [comment]
Thank you for your contribution!
